server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransactionLog.java

```diff
@@ -0,0 +1,385 @@
+  }
+
+  public int getNumTransactions() {
+    return this.log.getNumTransactions();
+  }
+
+  public List<DatafileTransaction> getTransactions() {
+    return this.log.getTransactions();
+  }
+
+  public boolean isEmpty() {
+    return this.log.isEmpty();
+  }
+
+  public Set<StoredTabletFile> getExpectedFiles() {
+    return this.log.getExpectedFiles();
+  }
+
+  public void compacted(Set<StoredTabletFile> files, Optional<StoredTabletFile> output) {
+    addTransaction(new DatafileTransaction.Compacted(files, output));
+  }
+
+  public void flushed(Optional<StoredTabletFile> newDatafile) {
+    addTransaction(new DatafileTransaction.Flushed(newDatafile));
+  }
+
+  public void bulkImported(StoredTabletFile file) {
+    addTransaction(new DatafileTransaction.BulkImported(file));
+  }
+
+  /**
+   * Add a transaction to the log. This will trim the size of the log if needed.
+   *
+   * @param transaction The transaction to add
+   */
+  private void addTransaction(DatafileTransaction transaction) {
+    this.log.setCapacity(getMaxSize());
+    this.log.addTransaction(transaction);
+  }
+
+  /**
+   * Get a string that provides a list of the transactions.
+   *
+   * @return a log dump
+   */
+  public String dumpLog() {
+    return this.log.dumpLog(extent, false);
+  }
+
+  public String dumpAndClearLog() {
+    return this.log.dumpLog(extent, true);
+  }
+
+  public void clearLog() {
+    this.log.clear();
+  }
+
+  public void resetLog(Set<StoredTabletFile> files) {
+    this.log.reset(files);
+  }
+
+  @Override
+  public String toString() {
+    return dumpLog();
+  }
+
+  /**
+   * A transaction log consists of the original file set and its timestamp, a set of transactions,
+   * and the final set of files after applying the transations. This class is immutable.
+   */
+  private static class TransactionLog {
+    private static final String DATE_FORMAT = "yyyyMMdd'T'HH:mm:ss.SSS";
+    private volatile long updateCount = 0;
+    // The time stamp of the initial file set
+    private volatile long initialTs;
+    // the initial file set
+    private StoredTabletFile[] initialFiles;
+    // the transactions
+    private Ring<DatafileTransaction> tabletLog;
+    // the final file set derived be applying the transactions to the initial file set
+    private StoredTabletFile[] finalFiles;
+
+    public TransactionLog(Set<StoredTabletFile> files) {
+      this(files.toArray(new StoredTabletFile[0]));
+    }
+
+    private TransactionLog(StoredTabletFile[] files) {
+      this(System.currentTimeMillis(), files, new Ring<>(0), files);
+    }
+
+    private TransactionLog(long initialTs, StoredTabletFile[] initialFiles,
+        Ring<DatafileTransaction> tabletLog, StoredTabletFile[] finalFiles) {
+      this.initialTs = initialTs;
+      this.initialFiles = initialFiles;
+      this.tabletLog = tabletLog;
+      this.finalFiles = finalFiles;
+      this.updateCount = tabletLog.getUpdateCount();
+    }
+
+    public void setCapacity(int maxSize) {
```

Comment:
I think this can be called by multiple threads concurrently.
```suggestion
    public synchronized void setCapacity(int maxSize) {
```

Reply:
No it can't,  each of those calls are within the Tablet lock.,